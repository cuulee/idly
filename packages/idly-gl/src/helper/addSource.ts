export function addSource(layer: any, source: string) {
  return {
    source,
    id: source + '-' + layer.id,
    ...layer
  };
}
