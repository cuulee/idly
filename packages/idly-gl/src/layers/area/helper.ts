import { PLUGIN_NAME } from '../../constants';

export const areaPaintStyle = {
  'line-color': '#551A8B',
  'line-width': {
    base: 4,
    stops: [[16, 4], [18, 40], [22, 20]]
  },
  'line-opacity': {
    base: 0.2,
    stops: [[16, 0.6], [18, 0.3], [22, 0.4]]
  },
  'line-offset': {
    base: 4,
    stops: [[16, 4], [18, 16], [22, 12]]
  },
  'line-blur': {
    base: 2,
    stops: [[16, 4], [18, 8], [22, 12]]
  }
};
