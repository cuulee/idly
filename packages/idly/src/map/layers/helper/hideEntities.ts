import { List, Set } from 'immutable';

import { ILayerSpec } from 'map/layers/layerFactory';
import { setSubtractEntities } from 'map/utils/setSubtract';
import { Entities } from 'osm/entities/entities';

export function isList(x: string | List<any>): x is List<any> {
  return List.isList(x);
}

const findNotInId = n => isList(n) && n.get(0) === '!in' && n.get(1) === 'id';

export function hideEntities(
  layer: ILayerSpec,
  oldEntities: Entities,
  newEntities: Entities
) {
  return layer.update('filter', filter => {
    const addedEntities = setSubtractEntities(newEntities, oldEntities);
    const index = filter.findIndex(findNotInId);
    if (addedEntities.size > 0) {
      if (index === -1) return filter;
      return filter.remove(index);
    }
    const removedEntities = setSubtractEntities(oldEntities, newEntities);
    if (removedEntities.size > 0) {
      const toRemove = removedEntities.map(e => e.id);
      const target = List(['!in', 'id']).concat(toRemove).toList();
      if (index === -1) {
        return filter.push(target);
      }

      return filter.update(index, (t: List<string>) => {
        return t.concat(toRemove).toOrderedSet().toList();
      });
    }
    return filter;
  });
}
