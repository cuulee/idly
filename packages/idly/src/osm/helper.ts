import * as _ from 'lodash';

export function osmIsInterestingTag(key) {
  return (
    key !== 'attribution' &&
    key !== 'created_by' &&
    key !== 'source' &&
    key !== 'odbl' &&
    key.indexOf('tiger:') !== 0
  );
}

// using WGS84 polar radius (6356752.314245179 m)
// const = 2 * PI * r / 360
export function geoMetersToLat(m) {
  return m / 110946.257617;
}

export function geoMetersToLon(m, atLat) {
  return Math.abs(atLat) >= 90
    ? 0
    : m / 111319.490793 / Math.abs(Math.cos(atLat * (Math.PI / 180)));
}
export function geoPolygonContainsPolygon(outer, inner) {
  return _.every(inner, function(point) {
    return geoPointInPolygon(point, outer);
  });
}
export function geoLineIntersection(a, b) {
  function subtractPoints(point1, point2) {
    return [point1[0] - point2[0], point1[1] - point2[1]];
  }
  function crossProduct(point1, point2) {
    return point1[0] * point2[1] - point1[1] * point2[0];
  }

  var p = [a[0][0], a[0][1]],
    p2 = [a[1][0], a[1][1]],
    q = [b[0][0], b[0][1]],
    q2 = [b[1][0], b[1][1]],
    r = subtractPoints(p2, p),
    s = subtractPoints(q2, q),
    uNumerator = crossProduct(subtractPoints(q, p), r),
    denominator = crossProduct(r, s);

  if (uNumerator && denominator) {
    var u = uNumerator / denominator,
      t = crossProduct(subtractPoints(q, p), s) / denominator;

    if (t >= 0 && t <= 1 && u >= 0 && u <= 1) {
      return geoInterp(p, p2, t);
    }
  }

  return null;
}
export function geoInterp(p1, p2, t) {
  return [p1[0] + (p2[0] - p1[0]) * t, p1[1] + (p2[1] - p1[1]) * t];
}

export function geoPolygonIntersectsPolygon(outer, inner, checkSegments) {
  function testSegments(outer, inner) {
    for (var i = 0; i < outer.length - 1; i++) {
      for (var j = 0; j < inner.length - 1; j++) {
        var a = [outer[i], outer[i + 1]],
          b = [inner[j], inner[j + 1]];
        if (geoLineIntersection(a, b)) return true;
      }
    }
    return false;
  }

  function testPoints(outer, inner) {
    return _.some(inner, function(point) {
      return geoPointInPolygon(point, outer);
    });
  }

  return (
    testPoints(outer, inner) || (!!checkSegments && testSegments(outer, inner))
  );
}
export function geoPointInPolygon(point, polygon) {
  var x = point[0],
    y = point[1],
    inside = false;

  for (var i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    var xi = polygon[i][0],
      yi = polygon[i][1];
    var xj = polygon[j][0],
      yj = polygon[j][1];

    var intersect =
      yi > y !== yj > y && x < (xj - xi) * (y - yi) / (yj - yi) + xi;
    if (intersect) inside = !inside;
  }

  return inside;
}

export const dataDeprecated = [];
export var osmOneWayTags = {
  aerialway: {
    chair_lift: true,
    mixed_lift: true,
    't-bar': true,
    'j-bar': true,
    platter: true,
    rope_tow: true,
    magic_carpet: true,
    yes: true
  },
  highway: {
    motorway: true,
    motorway_link: true
  },
  junction: {
    roundabout: true
  },
  man_made: {
    'piste:halfpipe': true
  },
  'piste:type': {
    downhill: true,
    sled: true,
    yes: true
  },
  waterway: {
    canal: true,
    ditch: true,
    drain: true,
    river: true,
    stream: true
  }
};

export function geoCross(o, a, b) {
  return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0]);
}

export function geoExtent(min, max): void {
  if (!(this instanceof geoExtent)) return new geoExtent(min, max);
  if (min instanceof geoExtent) {
    return min;
  } else if (
    min &&
    min.length === 2 &&
    min[0].length === 2 &&
    min[1].length === 2
  ) {
    this[0] = min[0];
    this[1] = min[1];
  } else {
    this[0] = min || [Infinity, Infinity];
    this[1] = max || min || [-Infinity, -Infinity];
  }
}

geoExtent.prototype = new Array(2);

_.extend(geoExtent.prototype, {
  equals: function(obj) {
    return (
      this[0][0] === obj[0][0] &&
      this[0][1] === obj[0][1] &&
      this[1][0] === obj[1][0] &&
      this[1][1] === obj[1][1]
    );
  },

  extend: function(obj) {
    if (!(obj instanceof geoExtent)) obj = new geoExtent(obj);
    return geoExtent(
      [Math.min(obj[0][0], this[0][0]), Math.min(obj[0][1], this[0][1])],
      [Math.max(obj[1][0], this[1][0]), Math.max(obj[1][1], this[1][1])]
    );
  },

  _extend: function(extent) {
    this[0][0] = Math.min(extent[0][0], this[0][0]);
    this[0][1] = Math.min(extent[0][1], this[0][1]);
    this[1][0] = Math.max(extent[1][0], this[1][0]);
    this[1][1] = Math.max(extent[1][1], this[1][1]);
  },

  area: function() {
    return Math.abs((this[1][0] - this[0][0]) * (this[1][1] - this[0][1]));
  },

  center: function() {
    return [(this[0][0] + this[1][0]) / 2, (this[0][1] + this[1][1]) / 2];
  },

  rectangle: function() {
    return [this[0][0], this[0][1], this[1][0], this[1][1]];
  },

  bbox: function() {
    return {
      minX: this[0][0],
      minY: this[0][1],
      maxX: this[1][0],
      maxY: this[1][1]
    };
  },

  polygon: function() {
    return [
      [this[0][0], this[0][1]],
      [this[0][0], this[1][1]],
      [this[1][0], this[1][1]],
      [this[1][0], this[0][1]],
      [this[0][0], this[0][1]]
    ];
  },

  contains: function(obj) {
    if (!(obj instanceof geoExtent)) obj = new geoExtent(obj);
    return (
      obj[0][0] >= this[0][0] &&
      obj[0][1] >= this[0][1] &&
      obj[1][0] <= this[1][0] &&
      obj[1][1] <= this[1][1]
    );
  },

  intersects: function(obj) {
    if (!(obj instanceof geoExtent)) obj = new geoExtent(obj);
    return (
      obj[0][0] <= this[1][0] &&
      obj[0][1] <= this[1][1] &&
      obj[1][0] >= this[0][0] &&
      obj[1][1] >= this[0][1]
    );
  },

  intersection: function(obj) {
    if (!this.intersects(obj)) return new geoExtent();
    return new geoExtent(
      [Math.max(obj[0][0], this[0][0]), Math.max(obj[0][1], this[0][1])],
      [Math.min(obj[1][0], this[1][0]), Math.min(obj[1][1], this[1][1])]
    );
  },

  percentContainedIn: function(obj) {
    if (!(obj instanceof geoExtent)) obj = new geoExtent(obj);
    var a1 = this.intersection(obj).area(),
      a2 = this.area();

    if (a1 === Infinity || a2 === Infinity || a1 === 0 || a2 === 0) {
      return 0;
    } else {
      return a1 / a2;
    }
  },

  padByMeters: function(meters) {
    var dLat = geoMetersToLat(meters),
      dLon = geoMetersToLon(meters, this.center()[1]);
    return geoExtent(
      [this[0][0] - dLon, this[0][1] - dLat],
      [this[1][0] + dLon, this[1][1] + dLat]
    );
  },

  toParam: function() {
    return this.rectangle().join(',');
  }
});

export function actionReverse(wayId, options) {
  var replacements = [
      [/:right$/, ':left'],
      [/:left$/, ':right'],
      [/:forward$/, ':backward'],
      [/:backward$/, ':forward']
    ],
    numeric = /^([+\-]?)(?=[\d.])/,
    roleReversals = {
      forward: 'backward',
      backward: 'forward',
      north: 'south',
      south: 'north',
      east: 'west',
      west: 'east'
    };

  function reverseKey(key) {
    for (var i = 0; i < replacements.length; ++i) {
      var replacement = replacements[i];
      if (replacement[0].test(key)) {
        return key.replace(replacement[0], replacement[1]);
      }
    }
    return key;
  }

  function reverseValue(key, value) {
    if (key === 'incline' && numeric.test(value)) {
      return value.replace(numeric, function(_, sign) {
        return sign === '-' ? '' : '-';
      });
    } else if (key === 'incline' || key === 'direction') {
      return { up: 'down', down: 'up' }[value] || value;
    } else if (options && options.reverseOneway && key === 'oneway') {
      return { yes: '-1', '1': '-1', '-1': 'yes' }[value] || value;
    } else {
      return { left: 'right', right: 'left' }[value] || value;
    }
  }

  function reverseDirectionTags(node) {
    // Update the direction based tags as appropriate then return an updated node
    return node.update({
      tags: _.transform(
        node.tags,
        function(acc, tagValue, tagKey) {
          // See if this is a direction tag and reverse (or use existing value if not recognised)
          if (tagKey === 'direction') {
            acc[tagKey] =
              {
                forward: 'backward',
                backward: 'forward',
                left: 'right',
                right: 'left'
              }[tagValue] || tagValue;
          } else {
            // Use the reverseKey method to cater for situations such as traffic_sign:forward=stop
            // This will pass through other tags unchanged
            acc[reverseKey(tagKey)] = tagValue;
          }
          return acc;
        },
        {}
      )
    });
  }

  function reverseTagsOnNodes(graph, nodeIds) {
    // Reverse the direction of appropriate tags attached to the nodes (#3076)
    return (
      _(nodeIds)
        // Get each node from the graph
        .map(function(nodeId) {
          return graph.entity(nodeId);
        })
        // Check tags on the node, if there aren't any, we can skip
        .filter(function(existingNode) {
          return existingNode.tags !== undefined;
        })
        // Get a new version of each node with the appropriate tags reversed
        .map(function(existingNode) {
          return reverseDirectionTags(existingNode);
        })
        // Chain together consecutive updates to the graph for each updated node and return
        .reduce(function(accGraph, value) {
          return accGraph.replace(value);
        }, graph)
    );
  }

  return function(graph) {
    var way = graph.entity(wayId),
      nodes = way.nodes.slice().reverse(),
      tags = {},
      key,
      role;

    for (key in way.tags) {
      tags[reverseKey(key)] = reverseValue(key, way.tags[key]);
    }

    graph.parentRelations(way).forEach(function(relation) {
      relation.members.forEach(function(member, index) {
        if (member.id === way.id && (role = roleReversals[member.role])) {
          relation = relation.updateMember({ role: role }, index);
          graph = graph.replace(relation);
        }
      });
    });

    // Reverse any associated directions on nodes on the way and then replace
    // the way itself with the reversed node ids and updated way tags
    return reverseTagsOnNodes(graph, nodes).replace(
      way.update({ nodes: nodes, tags: tags })
    );
  };
}
