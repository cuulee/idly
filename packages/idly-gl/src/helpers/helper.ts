import area from '@turf/area';
import bboxClip from '@turf/bbox-clip';
import bboxPolygon from '@turf/bbox-polygon';
import { FeatureCollection } from '@turf/helpers';
import { BBox, mercator, Tile } from 'idly-common/lib/geo';
import parser from 'idly-faster-osm-parser';
import { IDLY_NS } from '../constants';

export function addSource(layer: any, source: string) {
  return {
    ...layer,
    source,
    id: getNameSpacedLayerId(layer.id, source),
  };
}

export function getNameSpacedLayerId(layerId: string, source: string) {
  return source + IDLY_NS + layerId;
}

export function reverseGetNameSpacedLayerId(id: string) {
  return id.split(IDLY_NS)[1];
}

export function tilesFilterSmall(
  tiles: Tile[],
  viewportBbox: BBox,
  minimumOverlap: number
): Tile[] {
  const final = tiles
    .map(tile => {
      const tilePolygon = bboxPolygon(mercator.bbox(tile.x, tile.y, tile.z));
      const tileInsideBbox = bboxClip(tilePolygon, viewportBbox);

      const fractionVisible =
        area(tileInsideBbox) / area(bboxPolygon(viewportBbox));

      return { fractionVisible, tile };
    })
    .sort((a, b) => b.fractionVisible - a.fractionVisible)
    .reduce(
      (prev, { fractionVisible, tile }) => {
        if (prev.sum >= minimumOverlap) {
          return prev;
        }
        prev.tiles.push(tile);
        prev.sum += fractionVisible;

        return prev;
      },
      { sum: 0, tiles: [] as Tile[] }
    );
  console.log(
    'thresh',
    minimumOverlap,
    'orig',
    tiles.length,
    'removed',
    tiles.length - final.tiles.length
  );
  return final.tiles;
}

export function blankFC(): FeatureCollection<any, any> {
  return {
    type: 'FeatureCollection',
    features: [],
  };
}

const cache: any = new Map();

export async function fetchTileXml(
  x: number,
  y: number,
  zoom: number
): Promise<any> {
  // return parser(rsp);
  const bboxStr = mercator.bbox(x, y, zoom).join(',');
  if (cache.has(bboxStr)) {
    return cache.get(bboxStr);
  }
  const response = await fetch(
    `https://www.openstreetmap.org/api/0.6/map?bbox=${bboxStr}`
  );
  if (!response.ok) {
    throw new Error(response.statusText);
  }
  const entities = parser(await response.text());
  cache.set(bboxStr, entities);
  return entities;
}

export function bboxify(e: any, factor: number) {
  return [
    [e.point.x - factor, e.point.y - factor],
    [e.point.x + factor, e.point.y + factor],
  ];
}
export const hideVersion = (index?: string) =>
  index && modifiedIdParse(index)[0];

export const modifiedIdParse = (index: string) => index.split('#');

export const distance = (
  [x1, y1]: [number, number],
  [x2, y2]: [number, number]
) => Math.sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1));

export function tapLog<T>(fn: T) {
  if (typeof fn !== 'function') {
    console.log('Log-- expression', fn);
    return fn;
  }
  return (...args: any[]) => {
    console.log('Log-- input', ...args);
    const result = (fn as any).apply(null, args);
    console.log('Log-- output', result);
    return result;
  };
}

export function bindThis(target: any, key: any, { value: fn }: any) {
  return {
    configurable: true,
    get() {
      const value = fn.bind(this);
      Object.defineProperty(this, key, {
        value,
        configurable: true,
        writable: true,
      });
      return value;
    },
  };
}
