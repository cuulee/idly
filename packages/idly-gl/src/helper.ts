import area from '@turf/area';
import bboxClip from '@turf/bbox-clip';
import bboxPolygon from '@turf/bbox-polygon';
import { FeatureCollection } from '@turf/helpers';
import { BBox, mercator, Tile } from 'idly-common/lib/geo';
import parser from 'idly-faster-osm-parser';

export function addSource(layer: any, source: string) {
  return {
    ...layer,
    source,
    id: getNameSpacedLayerId(layer.id, source),
  };
}
export function getNameSpacedLayerId(layerId: string, source: string) {
  return source + '-' + layerId;
}

export function tilesFilterSmall(
  tiles: Tile[],
  viewportBbox: BBox,
  minimumOverlap: number
): Tile[] {
  const final = tiles.filter(tile => {
    const tilePolygon = bboxPolygon(mercator.bbox(tile.x, tile.y, tile.z));
    const tileInsideBbox = bboxClip(tilePolygon, viewportBbox);

    const fractionVisible = area(tileInsideBbox) / area(tilePolygon);

    return fractionVisible >= minimumOverlap;
  });

  // console.log('main');
  // console.log(
  //   JSON.stringify({
  //     type: 'FeatureCollection',
  //     features: [
  //       bboxPolygon(viewportBbox),
  //       ...final.map(tile =>
  //         bboxPolygon(mercator.bbox(tile.x, tile.y, tile.z))
  //       ),
  //     ],
  //   })
  // );
  // console.log('mainEnd');

  return final;
}

export function blankFC(): FeatureCollection<any, any> {
  return {
    type: 'FeatureCollection',
    features: [],
  };
}

const cache: any = new Map();

export async function fetchTileXml(
  x: number,
  y: number,
  zoom: number
): Promise<any> {
  const bboxStr = mercator.bbox(x, y, zoom).join(',');
  if (cache.has(bboxStr)) {
    return cache.get(bboxStr);
  }
  const response = await fetch(
    `https://www.openstreetmap.org/api/0.6/map?bbox=${bboxStr}`
  );
  if (!response.ok) {
    throw new Error(response.statusText);
  }
  const entities = parser(await response.text());
  cache.set(bboxStr, entities);
  return entities;
}

export function bboxify(e: any, factor: number) {
  return [
    [e.point.x - factor, e.point.y - factor],
    [e.point.x + factor, e.point.y + factor],
  ];
}

export const hideVersion = (index: string) =>
  index && modifiedIdParse(index)[0];

const modifiedIdParse = (index: string) => index.split('#');
