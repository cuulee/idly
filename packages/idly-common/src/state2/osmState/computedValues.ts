import { Entity, EntityType } from '../../osm/structures';
import { setCreate, setEqual } from '../helper';

export type DerivedTable = Map<string, Derived>;

export interface Derived {
  readonly entity: Entity;
  readonly parentWays: Set<string>;
  readonly parentRelations: Set<string>;
}

export const derivedCreate = (entity: Entity): Derived => ({
  entity,
  parentRelations: setCreate<string>(),
  parentWays: setCreate<string>(),
});

export function updateDerivedValues(
  entityTable: Map<string, Entity>,
  derivedTable: Map<string, Derived> = new Map()
): DerivedTable {
  const intermediateTable = computeParents(entityTable);

  const result = new Map();

  intermediateTable.forEach((compute, id) => {
    const prevVal = derivedTable.get(id);
    // both values are same, it is useless
    // to update the reference.
    if (isComputedEqual(compute, prevVal)) {
      return result.set(id, prevVal);
    }
    // updates the derivedTable for futureUse
    derivedTable.set(id, compute);
    return result.set(id, compute);
  });

  return result;
}

const isComputedEqual = (fresh: Derived, cB?: Derived) => {
  if (!fresh || !cB) {
    return false;
  }
  if (fresh.entity !== cB.entity) {
    return false;
  }
  if (fresh.entity.type === EntityType.NODE) {
    if (!setEqual(fresh.parentWays, cB.parentWays)) {
      return false;
    }
  }
  if (!setEqual(fresh.parentRelations, cB.parentRelations)) {
    return false;
  }
  return true;
};

export const computeParents = (
  entityTable: Map<string, Entity>
): DerivedTable => {
  const computedTable: Map<string, Derived> = new Map();

  entityTable.forEach(entity =>
    computedTable.set(entity.id, derivedCreate(entity))
  );

  computedTable.forEach(comput => {
    if (comput.entity.type === EntityType.WAY) {
      comput.entity.nodes.forEach(nodeId => {
        const nodeCompute = computedTable.get(nodeId);
        if (nodeCompute) {
          nodeCompute.parentWays.add(comput.entity.id);
        }
      });
    } else if (comput.entity.type === EntityType.RELATION) {
      comput.entity.members.forEach(({ id }) => {
        const entityComp = computedTable.get(id);
        if (entityComp) {
          entityComp.parentRelations.add(comput.entity.id);
        }
      });
    }
  });
  return computedTable;
};
