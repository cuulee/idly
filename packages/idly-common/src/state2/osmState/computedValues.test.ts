import {
  nodeFactory,
  relationFactory,
  wayFactory,
} from '../../osm/entityFactory';
import { setCreate } from '../helper';
import { updateDerivedValues } from './computedValues';

const n1 = nodeFactory({
  id: 'n1',
});
const n2 = nodeFactory({
  id: 'n2',
});
const n3 = nodeFactory({
  id: 'n3',
});
const n4 = nodeFactory({
  id: 'n4',
});
const n5 = nodeFactory({
  id: 'n5',
});

const n6 = nodeFactory({
  id: 'n6',
});

const n7 = nodeFactory({
  id: 'n7',
});

const w1 = wayFactory({
  id: 'w1',
  nodes: ['n1', 'n2'],
});

const w2 = wayFactory({
  id: 'w2',
  nodes: ['n1', 'n3'],
});

const w3 = wayFactory({
  id: 'w3',
  nodes: ['n3', 'n4', 'n5'],
});

const r1 = relationFactory({
  id: 'r1',
  members: [{ id: 'n4', ref: 'n4' }, { id: 'w2', ref: 'w2' }],
});

const r2 = relationFactory({
  id: 'r2',
  members: [
    { id: 'w1', ref: 'w1' },
    { id: 'w3', ref: 'w3' },
    { id: 'n6', ref: 'n6' },
  ],
});

const r3 = relationFactory({
  id: 'r3',
  members: [
    { id: 'r1', ref: 'r1' },
    { id: 'w2', ref: 'w2' },
    { id: 'n1', ref: 'n1' },
  ],
});
const mapFromObj = (o: any): Map<string, any> =>
  Object.keys(o).reduce((prev, k) => {
    prev.set(k, o[k]);
    return prev;
  }, new Map());

describe('computed values', () => {
  it('it creates even when entity doesnt exist', () => {
    const entityTable = mapFromObj({
      n1,
      r1,
      w1,
    });
    const computed: any = updateDerivedValues(entityTable);
    expect(computed.get('n1').parentWays).toEqual(setCreate(['w1']));

    expect(updateDerivedValues(entityTable)).toMatchSnapshot();
  });

  it('should mark parents correctly', () => {
    const entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w1,
      w2,
      w3,
    });
    const computed = updateDerivedValues(entityTable);

    const parentWays: any = {};
    const parentRelations: any = {};

    computed.forEach(c => {
      parentWays[c.entity.id] = c.parentWays;
      parentRelations[c.entity.id] = c.parentRelations;
    });

    expect(parentWays).toMatchSnapshot();
    expect(parentRelations).toMatchSnapshot();
  });

  it('should remove entity if not found in the new table ', () => {
    let entityTable: any = mapFromObj({
      n1,
      n2,
      n5,
      r1,
      w2,
    });

    const computed: any = updateDerivedValues(entityTable);

    expect(computed.get('n1').parentWays).toEqual(setCreate(['w2']));

    entityTable = mapFromObj({
      n1,
      n5,
      r1,
      w2,
    });

    const computed2 = updateDerivedValues(entityTable, computed);
    expect(computed2.get('n1')).toBe(computed.get('n1'));
    expect(computed2.get('n5')).toBe(computed.get('n5'));

    expect(computed2.get('n2')).toBe(undefined);

    entityTable = mapFromObj({
      n1,
      n5,
      r1,
      w2,
      w3,
    });

    const computed3 = updateDerivedValues(entityTable, computed2);
    expect(computed3.get('n1')).toBe(computed2.get('n1'));
    expect(computed3.get('w2')).toBe(computed.get('w2'));
    expect(computed3.get('n5')).not.toBe(computed.get('n5'));
    expect(computed3.get('n5')).toEqual({
      entity: n5,
      parentRelations: setCreate([]),
      parentWays: setCreate(['w3']),
    });

    const newRelation = relationFactory({
      id: 'r55',
      members: [n1, n5, r1, w2, w3].map(e => ({ id: e.id, ref: e.id })),
    });

    const computed4 = updateDerivedValues(
      mapFromObj({
        n1,
        n5,
        r1,
        r55: newRelation,
        w2,
        w3,
      }),
      computed3
    );

    const parentWays: any = {};
    const parentRelations: any = {};

    computed4.forEach(c => {
      parentWays[c.entity.id] = c.parentWays;
      parentRelations[c.entity.id] = c.parentRelations;
    });

    expect(parentWays).toMatchSnapshot();
    expect(parentRelations).toMatchSnapshot();

    const computed5 = updateDerivedValues(mapFromObj({}), computed3);

    expect(computed5.size).toBe(0);
  });

  it('keeps the reference same when not modified', () => {
    let entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w1,
      w2,
      w3,
    });
    const computed = updateDerivedValues(entityTable);

    entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w2,
      w3,
    });

    const computed2 = updateDerivedValues(entityTable, computed);

    expect((computed2.get('n1') as any).parentWays).toEqual(setCreate(['w2']));
    expect((computed2.get('n2') as any).parentWays).toEqual(setCreate());

    expect(computed2.get('n3') as any).toBe(computed.get('n3') as any);
    expect(computed2.get('n4') as any).toBe(computed.get('n4') as any);
    expect(computed2.get('n5') as any).toBe(computed.get('n5') as any);
    expect(computed2.get('n6') as any).toBe(computed.get('n6') as any);
    expect(computed2.get('n7') as any).toBe(computed.get('n7') as any);
    expect(computed2.get('r1') as any).toBe(computed.get('r1') as any);
    expect(computed2.get('r2') as any).toBe(computed.get('r2') as any);
    expect(computed2.get('r3') as any).toBe(computed.get('r3') as any);
    expect(computed2.get('w2') as any).toBe(computed.get('w2') as any);
    expect(computed2.get('w3') as any).toBe(computed.get('w3') as any);

    expect((computed2.get('n3') as any).parentWays).toBe(
      (computed.get('n3') as any).parentWays
    );
    expect((computed2.get('n4') as any).parentWays).toBe(
      (computed.get('n4') as any).parentWays
    );
    expect((computed2.get('n5') as any).parentWays).toBe(
      (computed.get('n5') as any).parentWays
    );
    expect((computed2.get('n6') as any).parentWays).toBe(
      (computed.get('n6') as any).parentWays
    );
    expect((computed2.get('n7') as any).parentWays).toBe(
      (computed.get('n7') as any).parentWays
    );
    expect((computed2.get('r1') as any).parentWays).toBe(
      (computed.get('r1') as any).parentWays
    );
    expect((computed2.get('r2') as any).parentWays).toBe(
      (computed.get('r2') as any).parentWays
    );
    expect((computed2.get('r3') as any).parentWays).toBe(
      (computed.get('r3') as any).parentWays
    );
    expect((computed2.get('w2') as any).parentWays).toBe(
      (computed.get('w2') as any).parentWays
    );
    expect((computed2.get('w3') as any).parentWays).toBe(
      (computed.get('w3') as any).parentWays
    );
  });

  describe('it updates the derived table', () => {
    let entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w1,
      w2,
      w3,
    });
    const derivedTable = updateDerivedValues(entityTable);

    const w1Ref = derivedTable.get('w1');
    const r1Ref = derivedTable.get('r1');

    const n1Ref = derivedTable.get('n1');
    const n3Ref = derivedTable.get('n3');
    const n4Ref = derivedTable.get('n4');
    const n5Ref = derivedTable.get('n5');

    entityTable = mapFromObj({
      n1,
    });

    const compute1 = updateDerivedValues(entityTable, derivedTable);

    const n1Ref2 = compute1.get('n1');
    expect(compute1.size).toBe(1);
    expect(compute1.get('n1')).toMatchSnapshot();

    it('updates the derived table when n1 is modified', () => {
      expect(derivedTable.get('n1')).not.toBe(n1Ref);
    });

    it('keeps other elements in derived table intact', () => {
      expect(derivedTable.get('w1')).toBe(w1Ref);
      expect(derivedTable.get('n3')).toBe(n3Ref);
    });

    const compute2 = updateDerivedValues(
      mapFromObj({
        n1,
        n2,
        n3,
        n4,
        n5,
        n6,
        n7,
        r1,
        r2,
        r3,
        w1,
        w2,
        w3,
      }),
      derivedTable
    );
    it('changes the n1 ref ', () => {
      expect(derivedTable.get('n1')).not.toBe(n1Ref);
      expect(derivedTable.get('n1')).not.toBe(n1Ref2);
    });

    it('reverts n1 value back to the original', () => {
      expect(derivedTable.get('n1')).not.toEqual(n1Ref2);
      expect(derivedTable.get('n1')).toEqual(n1Ref);
    });

    it('keeps other elements in derived table intact', () => {
      expect(derivedTable.get('w1')).toBe(w1Ref);
      expect(derivedTable.get('n3')).toBe(n3Ref);
      expect(derivedTable.get('n4')).toBe(n4Ref);
      expect(derivedTable.get('n5')).toBe(n5Ref);
      expect(derivedTable.get('r1')).toBe(r1Ref);
    });
  });
});
