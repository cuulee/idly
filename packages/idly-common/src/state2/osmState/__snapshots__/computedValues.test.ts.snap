// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "entity": Object {
    "attributes": Object {
      "changeset": undefined,
      "timestamp": undefined,
      "uid": undefined,
      "user": undefined,
      "version": undefined,
      "visible": undefined,
    },
    "id": "n1",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {},
    "type": "node",
  },
  "parentRelations": Set {},
  "parentWays": Set {},
}
`;

exports[`computed values it creates even when entity doesnt exist 1`] = `
Map {
  "n1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
    },
  },
  "r1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "r1",
      "members": Array [
        Object {
          "id": "n4",
          "ref": "n4",
        },
        Object {
          "id": "w2",
          "ref": "w2",
        },
      ],
      "tags": Object {},
      "type": "relation",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "w1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
}
`;

exports[`computed values should mark parents correctly 1`] = `
Object {
  "n1": Set {
    "w1",
    "w2",
  },
  "n2": Set {
    "w1",
  },
  "n3": Set {
    "w2",
    "w3",
  },
  "n4": Set {
    "w3",
  },
  "n5": Set {
    "w3",
  },
  "n6": Set {},
  "n7": Set {},
  "r1": Set {},
  "r2": Set {},
  "r3": Set {},
  "w1": Set {},
  "w2": Set {},
  "w3": Set {},
}
`;

exports[`computed values should mark parents correctly 2`] = `
Object {
  "n1": Set {
    "r3",
  },
  "n2": Set {},
  "n3": Set {},
  "n4": Set {
    "r1",
  },
  "n5": Set {},
  "n6": Set {
    "r2",
  },
  "n7": Set {},
  "r1": Set {
    "r3",
  },
  "r2": Set {},
  "r3": Set {},
  "w1": Set {
    "r2",
  },
  "w2": Set {
    "r1",
    "r3",
  },
  "w3": Set {
    "r2",
  },
}
`;

exports[`computed values should remove entity if not found in the new table  1`] = `
Object {
  "n1": Set {
    "w2",
  },
  "n5": Set {
    "w3",
  },
  "r1": Set {},
  "r55": Set {},
  "w2": Set {},
  "w3": Set {},
}
`;

exports[`computed values should remove entity if not found in the new table  2`] = `
Object {
  "n1": Set {
    "r55",
  },
  "n5": Set {
    "r55",
  },
  "r1": Set {
    "r55",
  },
  "r55": Set {},
  "w2": Set {
    "r1",
    "r55",
  },
  "w3": Set {
    "r55",
  },
}
`;
